046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,86 @@
+ 
+KafkaAgent currently collects the `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is `3`(`RUNNING`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation
+
+To address the issues with restarts and log recovery, KafkaRoller needs to be aware of the current broker state.  If a broker is performing log recovery, KafkaRoller should fail the reconciliation instead of force restarting it.
+
+## Proposal
+ 
+The objective of this proposal is to introduce a REST API endpoint within KafkaAgent, which exposes metrics from a Kafka broker. Furthermore, KafkaRoller would be modified so that it can query this endpoint to determine the current state of a broker, thus preventing unnecessary restarts. Additionally, KafkaRoller would also display the remaining logs and segments to recover, allowing a human operator to observe the recovery process and determine if it is making progress or is stuck.
+
+### Implementation details
+
+*KafkaAgent*:
+- Add a web server to accept SSL connections on port 8443.
+- Use Jetty for the web server since Kafka already uses it therefore avoids an additional dependency.
+- Reuse the broker's own certificate that would be in the container for TLS authentication. 
+- Configure the server to require TLS client authentication. 
+- Expose an endpoint `v1/brokermetrics` that includes the API version in the URI path.
+- Extend KafkaAgent to collect `kafka.log:type=LogManager,name=remainingLogsToRecover` and `kafka.log:type=LogManager,name=remainingSegmentsToRecover` metrics.
+- Take the metric name as a parameter in the request query e.g `v1/brokermetrics?name=BrokerState`. Use the value of the MBean's `name` field for the request  e.g. `BrokerState` and  `remainingLogsToRecover`.
+- Use Jetty's native ServerHandler API to handle a `GET` request and return 
+    - an integer value for `BrokerState`.
+    - an integer value for  `remainingLogsToRecover`.
+    - an integer value for `remainingSegmentsToRecover`.
+- If no metric name was provided in the request or the requested metric name is not supported, return HTTP 404 (Not Found) to the client.
+
+*KafkaRoller*:
+- Use the Pod DNS name to query the KafkaAgent API.
+- Use the cluster operator's client certificate and keys to authenticate with the KafkaAgent API server.
+- Instead of checking the broker's readiness through `isReady()` function, utilise the new RESTful endpoint to determine if the broker is ready.
+- Include an additional field in the status section of the Kafka CR. This will hold data that will be utilised by KafkaRoller in case the reconciliation process is repeated while the log recovery process is still ongoing:
+```
+status:
+    <kafka_pod_name>:
+      restartTimeStamp: 
+      logRecoveryExceededTimeout: <true|false>
```

Comment:
If we wanna stick with this (see the comment below - I'm not sure I follow this), then there should be definitely something between `.status` and `.<kafka-pod-name>` ... the pod names should not just randomly mix into the status root. You will also not be able to model it because you would need `Map<String, Object>` for this and the status API has a given structure.

Reply:
+1 to Jakubs point. Also be explicit about _when_ this is present (i.e. I assume it's present in the CR `state` only for the duration of a pod having been restarted and not observed to have progressed to RUNNING broker state, so normally this would be missing from the `status` entirely?)

Reply:
As agreed above, we are not storing anything in the CR. I will remove this part